core/trino-main/src/main/java/io/trino/execution/scheduler/BatchTaskScheduler.java

```diff
@@ -0,0 +1,569 @@
+
+            Multimap<PlanFragmentId, ShufflePartitionHandle> shuffleInputs = inputPartitionHandleFutures.entrySet().stream()
+                    .collect(flatteningToImmutableListMultimap(Map.Entry::getKey, entry -> getFutureValue(entry.getValue()).stream()));
+
+            taskSource = batchTaskSourceFactory.create(
+                    session,
+                    stageExecution.getFragment(),
+                    shuffleInputs,
+                    stageExecution::recordGetSplitTime,
+                    inputBucketToPartitionMap,
+                    inputBucketNodeMap);
+        }
+
+        List<ListenableFuture<Void>> inputReadyBlockedFutures = inputShuffles.values().stream()
+                .map(Shuffle::isInputReady)
+                .filter(future -> !future.isDone())
+                .map(MoreFutures::toListenableFuture)
+                .map(MoreFutures::asVoid)
+                .collect(toImmutableList());
+
+        if (!inputReadyBlockedFutures.isEmpty()) {
+            blocked = asVoid(allAsList(inputReadyBlockedFutures));
+            return;
+        }
+
+        while (!queuedPartitions.isEmpty() || !taskSource.isFinished()) {
+            if (finishedPartitions.contains(queuedPartitions.peek())) {
+                queuedPartitions.poll();
+                continue;
+            }
+
+            while (queuedPartitions.isEmpty() && !taskSource.isFinished()) {
+                List<BatchTask> tasks = taskSource.getMoreTasks();
+                for (BatchTask task : tasks) {
+                    queuedPartitions.add(task.getPartitionId());
+                    partitionToTaskDescriptorMap.put(task.getPartitionId(), task);
+                }
+                if (taskSource.isFinished()) {
+                    outputShuffle.ifPresent(shuffle -> shuffle.noMoreOutputTasks(partitionToTaskDescriptorMap.keySet()));
+                }
+            }
+
+            if (queuedPartitions.isEmpty()) {
+                break;
+            }
+
+            int partition = queuedPartitions.peek();
+            BatchTask batchTask = requireNonNull(partitionToTaskDescriptorMap.get(partition), () -> "task descriptor missing for partition: %s" + partition);
+
+            if (acquireNodeFuture == null) {
+                acquireNodeFuture = nodeAllocator.acquire(new NodeRequirements(taskSource.getTaskCatalog(), batchTask.getHostsRequirement()));
+            }
+            if (!acquireNodeFuture.isDone()) {
+                blocked = asVoid(acquireNodeFuture);
+                return;
+            }
+            InternalNode node = getFutureValue(acquireNodeFuture);
+            acquireNodeFuture = null;
+
+            queuedPartitions.poll();
+
+            Multimap<PlanNodeId, Split> tableScanSplits = batchTask.getSplits();
+            Multimap<PlanNodeId, Split> remoteSplits = createRemoteSplits(batchTask.getShuffleInputs());
+
+            Multimap<PlanNodeId, Split> taskSplits = ImmutableListMultimap.<PlanNodeId, Split>builder()
```

Comment:
In case of bucketed or collocated join it can read both, a table (thus the splits) and a remote exchange (if one of the tables is not bucketed and must be repartitioned)